File: AmaroTest/app/src/main/java/com/test/amaro/amarotest/presentation/main/MainAdapter.java
package com.test.amaro.amarotest.presentation.main;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView.Adapter;
import android.support.v7.widget.RecyclerView.ViewHolder;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ImageView.ScaleType;
import android.widget.TextView;
import com.squareup.picasso.Picasso;
import com.test.amaro.amarotest.R;
import com.test.amaro.amarotest.models.Product;
import com.test.amaro.amarotest.presentation.main.MainActivity.OnProductClickListener;
import com.test.amaro.amarotest.presentation.main.MainAdapter.ProductViewHolder;
import java.util.ArrayList;
import java.util.List;

public final class MainAdapter extends Adapter<ProductViewHolder> {

    private final List<Product> products;
    private final Context context;
    private final OnProductClickListener itemClickListener;

    public final void setProducts(final List<Product> list) {
        this.products.clear();
        this.products.addAll(list);
        this.notifyDataSetChanged();
    }

    public int getItemCount() {
        return this.products.size();
    }

    public MainAdapter.ProductViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.item_product, parent, false);
        return new MainAdapter.ProductViewHolder(view, this.itemClickListener);
    }

    public void onBindViewHolder(MainAdapter.ProductViewHolder holder, int position) {
        holder.bind(this.products.get(position));
    }

    MainAdapter(final Context context, final OnProductClickListener itemClickListener) {
        this.context = context;
        this.itemClickListener = itemClickListener;
        this.products = new ArrayList<>();
    }

    final class ProductViewHolder extends ViewHolder {

        private final OnProductClickListener itemClickListener;

        final void bind(final Product item) {

            final TextView txtName = this.itemView.findViewById(R.id.text_name);
            final TextView txtPrice = this.itemView.findViewById(R.id.text_price);
            final ImageView image = this.itemView.findViewById(R.id.image);

            txtName.setText(item.getName());
            txtPrice.setText(item.getActualPrice());
            setProductImage(item, image);

            this.itemView.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    itemClickListener.onProductClick(item);
                }
            });
        }

        private void setProductImage(Product item, ImageView image) {

            if (item.getImage().length() > 0) {
                final Drawable imagePlaceHolderDrawable = ContextCompat
                        .getDrawable(context, R.drawable.ic_account_circle_gray_100dp);
                Picasso.with(context).load(item.getImage())
                        .placeholder(imagePlaceHolderDrawable)
                        .into(image);
            } else {
                image.setScaleType(ScaleType.CENTER_INSIDE);
                image.setImageDrawable(
                        ContextCompat.getDrawable(context, R.drawable.ic_cloud_off_grey_100dp));
            }
        }

        ProductViewHolder(View view, OnProductClickListener itemClickListener) {

            super(view);
            this.itemClickListener = itemClickListener;
        }
    }
}

File: AmaroTest/app/src/main/java/com/test/amaro/amarotest/main/MainAdapter.kt
package com.test.amaro.amarotest.main

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.squareup.picasso.Picasso
import com.test.amaro.amarotest.R
import com.test.amaro.amarotest.models.Product
import kotlinx.android.synthetic.main.product.view.*

class MainAdapter(val context: Context, private val itemClick: (Product) -> Unit) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val products: MutableList<Product> = mutableListOf()

    fun setProducts(list: List<Product>) {
        products.addAll(list)
        notifyItemRangeChanged(itemCount, products.size)
    }

    override fun getItemCount() = products.size

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = ProductViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.product, parent, false), itemClick)

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val viewHolder = holder as ProductViewHolder
        viewHolder.bind(products[position])
    }

    inner class ProductViewHolder(view: View, private val itemClick: (Product) -> Unit) : RecyclerView.ViewHolder(view) {

        fun bind(item: Product) {
            with(item) {

                itemView.text_name.text = item.name
                itemView.text_price.text = item.actualPrice
                itemView.setOnClickListener { itemClick(this) }

                item.image?.let {
                    if (it.isNotEmpty()) {
                        Picasso.with(context)
                            .load(it)
                            .placeholder(R.drawable.ic_account_circle_gray_24dp)
                            .into(itemView.image)
                    }
                }

            }
        }
    }
}
